drivers/wasapi/audio_driver_wasapi.cpp

```diff
@@ -463,36 +421,160 @@ Error AudioDriverWASAPI::audio_device_init(AudioDeviceWASAPI *p_device, bool p_i
+Error AudioDriverWASAPI::audio_device_init(AudioDeviceWASAPI *p_device, bool p_input, bool p_reinit, bool p_no_audio_client_3) {
+	// This function can be called recursively, so clean up before starting:
+	audio_device_finish(p_device);
+
+	// 1. Find the correct audio device
+	ComPtr<IMMDevice> output_device = find_output_device(p_device, p_input);
+	if (!output_device) {
+		if (p_reinit) {
+			return ERR_CANT_OPEN;
+		} else {
+			ERR_FAIL_COND_V_MSG(false, ERR_CANT_OPEN, "WASAPI: Failed to find audio device");
+		}
+	}
+
+	// 2. Register notification callback
+	ComPtr<IMMDeviceEnumerator> enumerator = nullptr;
+	HRESULT hr = CoCreateInstance(CLSID_MMDeviceEnumerator, nullptr, CLSCTX_ALL, IID_IMMDeviceEnumerator, (void **)&enumerator);
+	ERR_FAIL_COND_V(hr != S_OK, ERR_CANT_OPEN);
+
+	register_notification_callback(enumerator);
+
+	// 3. Determine if we should use IAudioClient3
+	using_audio_client_3 = !p_input; // IID_IAudioClient3 is only used for adjustable output latency (not input)
+	if (p_no_audio_client_3) {
+		using_audio_client_3 = false;
+	}
+
+	// 4. Activate audio client
+	hr = activate_audio_client(output_device, using_audio_client_3, &p_device->audio_client);
+	if (hr == S_FALSE) {
+		// Fall back to IAudioClient
+		using_audio_client_3 = false;
+		hr = activate_audio_client(output_device, false, &p_device->audio_client);
+	}
+
+	if (p_reinit) {
+		if (hr != S_OK) {
+			return ERR_CANT_OPEN;
+		}
+	} else {
+		ERR_FAIL_COND_V(hr != S_OK, ERR_CANT_OPEN);
+	}
+
+	// 5. Set client properties for IAudioClient3
+	if (using_audio_client_3) {
+		AudioClientProperties audioProps{};
+		audioProps.cbSize = sizeof(AudioClientProperties);
+		audioProps.bIsOffload = FALSE;
+		audioProps.eCategory = AudioCategory_GameEffects;
+
+		hr = ((IAudioClient3 *)p_device->audio_client)->SetClientProperties(&audioProps);
+		ERR_FAIL_COND_V_MSG(hr != S_OK, ERR_CANT_OPEN, "WASAPI: SetClientProperties failed with error 0x" + String::num_uint64(hr, 16) + ".");
+	}
+
+	// 6. Get and validate mix format
+	bool used_closest = false;
+	WAVEFORMATEX *pwfex = get_and_validate_mix_format(p_device->audio_client, used_closest);
+	if (!pwfex) {
+		return ERR_CANT_OPEN;
+	}
+
+	// 7. Validate and set format
+	if (!validate_and_set_format(p_device, pwfex)) {
+		CoTaskMemFree(pwfex);
+		return ERR_CANT_OPEN;
+	}
+
+	// 8. Initialize audio client
+	hr = initialize_audio_client(p_device, pwfex, p_input, using_audio_client_3);
+	if (hr == S_FALSE) {
+		// Fall back to IAudioClient
+		using_audio_client_3 = false;
+		hr = initialize_audio_client(p_device, pwfex, p_input, false);
+	}
+
+	if (p_reinit) {
+		if (hr != S_OK) {
+			CoTaskMemFree(pwfex);
+			return ERR_CANT_OPEN;
+		}
+	} else {
+		if (unlikely(hr != S_OK)) {
+			CoTaskMemFree(pwfex);
+			ERR_FAIL_V_MSG(ERR_CANT_OPEN, "WASAPI: Initialize failed with error 0x" + String::num_uint64(hr, 16) + ".");
+		}
+	}
+
+	// 9. Setup buffer and latency
+	setup_buffer_and_latency(p_device, p_input, using_audio_client_3, pwfex);
```

Comment:
Lost ability to check for failures on this line.
Missing boolean or HRESULT return.